### PR TITLE
Change cache getting and setting

### PR DESCRIPTION
**Pull request by rob-b**
The gateway api caches requests made to the other internal api endpoints
but by default cachecontrol caches these requests by their url. We want
to include a user's auth token in the cache key (for private requests)
so that we can cache by user.

When reading from the cache we now always use the authorization header
to build the cache key in conjunction with the request.url. When
actually reading from the backend store we rely on redis' mget method to
request both versions of the key (with and without the hash of the auth
header) and return the first value we find. We are biased towards keys
that do NOT include the auth hash but in practice public/private should
be a binary switch and so there should never be an issue of a less
preferred key being returned.

When writing to the cache we examine the response when creating a cache
key to determine if the response is private or public so we can set an
appropriate cache key

https://thephysicalnetwork.atlassian.net/browse/RO-586

### CODE REVIEW

**Comment by rob-b** [comment]
@kirberich I recall you mentioning something about not needing to check the vary header and to use a method but I can't remember what that method was. Something makes me thing you said to use `is_secure` but obv that's not it (since that's a django thing for checking if the request used tls).

File: cachecontrol/controller.py [hunk 1]
```diff
@@ -58,8 +59,28 @@ def _urlnorm(cls, uri):
...
-        return cls._urlnorm(uri)
+    def _authorization_digest(cls, authorization_header):
+        """Creates a hash digest of a request's authorization header.
+        """
+        if not authorization_header:
+            return ''
+        return hashlib.sha224(authorization_header.encode('utf-8')).hexdigest()
+
+    @classmethod
+    def cache_key(cls, request, response=None):
```
Comment (kirberich) [inline cachecontrol/controller.py]: I don't actually think we need the response (or even an `is_public` flag) anymore because we're using mget.

We can always just create both keys and return which ever one we find in the cache

File: cachecontrol/controller.py [hunk 2]
```diff
@@ -58,8 +59,28 @@ def _urlnorm(cls, uri):
...
+        """Create a cache_key from a url and request headers.
+
+        Requests are generally stored by their URL.
+        For requests that have an Authorization header, we add a hash of that header to the cache key.
+        The hash is only used to make sure we allow caching for each user - it's not a security feature.
+        """
+        normalized_url = cls._urlnorm(request.url)
+        authorization_hash = cls._authorization_digest(request.headers.get('Authorization'))
+        vary_values = response.headers.get('vary', '') if response and response.headers else ''
```
Comment (kirberich) [inline cachecontrol/controller.py]: See above, I think this can all go away - also it looks a bit weird that we first create the hash, then check the response headers and then potentially remove the hash again, but this will go away anyway

File: cachecontrol/controller.py [hunk 3]
```diff
@@ -90,21 +111,21 @@ def cached_request(self, request):
...
-        logger.debug('Looking up "%s" in the cache', cache_url)
         cc = self.parse_cache_control(request.headers)
 
         # Bail out if the request insists on fresh data
         if 'no-cache' in cc:
             logger.debug('Request header has "no-cache", cache bypassed')
             return False
 
-        if 'max-age' in cc and cc['max-age'] == 0:
+        if 'max-age' in cc and cc['max-age'] == '0':
```
Comment (kirberich) [inline cachecontrol/controller.py]: I think this might be nicer as `int(cc['max-age']) == 0`, just in case we actually do get back an integer for some reason

File: cachecontrol/controller.py [hunk 4]
```diff
@@ -254,9 +275,9 @@ def cache_response(self, request, response, body=None,
 
         cc_req = self.parse_cache_control(request.headers)
         cc = self.parse_cache_control(response_headers)
-
-        cache_url = self.cache_url(request.url)
-        logger.debug('Updating cache with response from "%s"', cache_url)
+        cache_key = self.cache_key(request)
+        write_cache_key = self.cache_key(request, response)
```
Comment (kirberich) [inline cachecontrol/controller.py]: again as above this can probably go away if we just always create the key with authentication

File: cachecontrol/controller.py [hunk 5]
```diff
@@ -317,11 +338,12 @@ def update_cached_response(self, request, response):
         This should only ever be called when we've sent an ETag and
         gotten a 304 as the response.
         """
-        cache_url = self.cache_url(request.url)
-
+        cache_key = self.cache_key(request)
+        write_cache_key = self.cache_key(request, response)
```
Comment (kirberich) [inline cachecontrol/controller.py]: And again here

**Review by kirberich — CHANGES_REQUESTED** [review]
This looks great! Really just one note on that we can probably get rid of the response in the cache_key method now as we're doing multi-get so it doesn't hurt to look for the auth-key variant even for public requests

File: cachecontrol/controller.py (hunk 3 above)
Comment (rob-b) [reply cachecontrol/controller.py]: Before fixing this bug I checked how they implement the header parsing and they don't do anything to cast to ints so unless we modify that ourselves only strings will ever come back. That said, max-age is not accessed consistently - it's a combination of this string/int comparison mistake, trying to cast to int and catching the error and an explicit `isdigit` check so I'll make a sweep now and handle it consistently everywhere

**Comment by rob-b** [comment]
@kirberich can you take another look?

File: tests/test_storage_redis.py [hunk 6]
```diff
@@ -2,14 +2,61 @@
...
+    def setup(self):
+        self.cache = RedisCache(fakeredis.FakeStrictRedis())
+
+    def teardown(self):
+        self.cache.clear()
+
+    @pytest.mark.parametrize("key, value, expected", [
+        ("key-a", 100, b'100'),
+        ("key-b", 948, b'948'),
+        ("key-b;deadbeef", 3, b'3'),
```
Comment (dotsbb) [inline tests/test_storage_redis.py:31]: Please add a test case with `value` `0` or empty list/string.

File: cachecontrol/adapter.py [hunk 7]
```diff
@@ -117,8 +117,8 @@ def _update_chunk_length(self):
 
         # See if we should invalidate the cache.
         if request.method in self.invalidating_methods and resp.ok:
-            cache_url = self.controller.cache_url(request.url)
-            self.cache.delete(cache_url)
+            cache_key = self.controller.cache_key(request)
```
Comment (kkorekk) [inline cachecontrol/adapter.py:120]: `self.invalidating_methods` shouldn't be extended with `PATCH` ? We're using this method for updating user profile.

File: cachecontrol/caches/redis_cache.py [hunk 8]
```diff
@@ -20,7 +20,13 @@ def __init__(self, conn):
...
 
     def get(self, key):
-        return self.conn.get(key)
+        # a key might be of the format foo;bar (where `;` is our default delimter).
+        # To get this key's value from redis we do `mget foo foo;bar` (this
+        # returns a list of values for each key searched for) and return the
+        # first match
+        key_prefix, _ = keybreaker(key)
+        key = [key_prefix, key]
+        return next((i for i in self.conn.mget(key) if i), None)
```
Comment (kkorekk) [inline cachecontrol/caches/redis_cache.py]: I think `next` is not required here, `mget` will return None if key doesn't exist so this can be simplified to
```python
values = self.conn.mget([key_prefix, key])
return values[0] or values[1]
```

**Comment by rob-b** [comment]
If `mget` returns `None` then

values = self.conn.mget([key_prefix, key])return values[0] or values[1]


Would be an `NoneType is not scriptable` error or a type error (not sure
which). Either way, I think "simpler" is just a matter of taste here

**Comment by kkorekk** [comment]
@rob-b `mget` returns a list with `None` in place of missing keys, but this is minor comment, so you can ignore it.

**Comment by rob-b** [comment]
@kkorekk @dotsbb I just rebased this branch with a few changes; can you take another look.

* added `PATCH` to `invalidating_methods`
* added some tests for `False`-y values

**Review by dotsbb — APPROVED** [review]

**Review by kkorekk — APPROVED** [review]

**Comment by olasitarska** [comment]
@kirberich are you happy for this to be merged?

**Review by kirberich — APPROVED** [review]
